### ISSUE TITLE
[tmva][sofie] Improve numerical stability and overflow protection for Softplus operator

### ISSUE THREAD

**Issue by AdityaDRathore** [root-project/root#21071]
### Explain what you would like to see improved and how.

The current Softplus implementation (recently merged via #21050) relies on the `ROperator_BasicUnary` trait with the standard formula `std::log(std::exp(x) + 1)`. While functionally correct for small inputs, this implementation has two significant numerical limitations in edge cases for scientific computing:

* **Overflow:** For inputs $x > 88.72$ (in float32), `std::exp(x)` overflows to infinity, causing the operator to return `inf`. The expected behavior for large inputs is linear ($f(x) \approx x$).
* **Precision Loss:** For large negative inputs, `exp(x)` becomes very small. Computing `1 + exp(x)` discards significant digits before the log calculation, leading to precision errors.

Leading inference frameworks (PyTorch, TensorFlow) mitigate this by using a threshold for the linear region and `log1p` for the curve.

## Proposal
I propose updating the Softplus logic to use numerical stability safeguards:

1.  **Thresholding:** For $x \ge 20.0f$, return $x$.
    * At $x=20$, $\ln(1+e^{20}) \approx 20.000000002$, which is indistinguishable from $20.0$ in float32.
    * This prevents the `exp()` overflow completely for large positive values.
2.  **Precision:** For $x < 20.0f$, use `std::log1p(std::exp(x))`.
    * `std::log1p(z)` computes $\ln(1+z)$ accurately even when $z \approx 0$, preserving precision for negative inputs.

### Proposed Logic:
```cpp
// Threshold 20.0f defined as exact hexfloat 0x1.4p+4f
return (x >= 20.0f) ? x : std::log1p(std::exp(x));
```
## Action Plan
I have a complete implementation ready (from PR #21023) that includes:

* **Bit-exact Hexfloat constants** for the threshold.
* **Comprehensive GTest suite** (`TestSofieSoftplus.cxx`) verifying:
    * **Overflow protection** (inputs > 89 return $x$, not `Inf`).
    * **Stability** in the linear region.
    * **Correct usage** of `std::log1p` vs `std::log`.

I can submit a Pull Request to patch the `ROperator_BasicUnary` implementation to include this logic and integrate the validation suite.

### ROOT version

master 

### Installation method

build from source

### Operating system

Linux (Ubuntu 24.04)

### Additional context

**References & Prior Art:**
* **PyTorch Standard:** PyTorch's `Softplus` implementation utilizes a `threshold` parameter (default 20) to revert to a linear function for large inputs to avoid overflow.
  * *Source:* https://pytorch.org/docs/stable/generated/torch.nn.Softplus.html
* **TensorFlow:** Similarly uses `softplus(x) = x` for `x > threshold`.

**Related PRs:**
* **Current Implementation:** #21050 (Merged) - Established the operator.
* **Proposed Code Source:** #21023 (Closed) - I have a fully tested branch with the `log1p` and `threshold` logic ready to port.

cc: @lmoneta @bellenot @sanjibansg

**Comment by sanjibansg** [comment]
Hi @AdityaDRathore ,
Thanks for creating this issue. I agree to the improvements you proposed. It would be great if you make changes to your PR #21023 where you add such checks within the `ROperator_BasicUnary` file.

Thanks,
Sanjiban

**Comment by AdityaDRathore** [comment]
> Hi [@AdityaDRathore](https://github.com/AdityaDRathore) , Thanks for creating this issue. I agree to the improvements you proposed. It would be great if you make changes to your PR [#21023](https://github.com/root-project/root/pull/21023) where you add such checks within the `ROperator_BasicUnary` file.
> 
> Thanks, Sanjiban

Understood @sanjibansg. I will reopen PR #21023, rebase it on the latest master  and implement the stability logic directly inside ROperator_BasicUnary.hxx via the traits specialization. I'll also include the comprehensive GTest suite. Updating the PR now

**Comment by AdityaDRathore** [comment]
Hi @sanjibansg,

As discussed, I have submitted PR #21092 which implements the stability fix directly into `ROperator_BasicUnary`
It includes:
- Hexfloat thresholding (20.0f) for overflow protection.
- l`og1p` stabilization for precision.
- `TestSofieSoftplus` validation suite.

Thanks,
Aditya

**Comment by dpiparo** [comment]
Could this item be re-opened on the SOFIE dev repository?  In any case we want the fix to be applied there and then ported to ROOT as a patch afterwards.

**Comment by AdityaDRathore** [comment]
> Could this item be re-opened on the SOFIE dev repository? In any case we want the fix to be applied there and then ported to ROOT as a patch afterwards.

Hi @dpiparo,

Thank you for the guidance. I will re-open both issues on the SOFIE dev repository and submit the fixes there. I'll also port my PyTorch parser PR (#21528) and the Softplus stability fix (#21092) to target ML4EP/SOFIE.

Should the corresponding PRs on root-project/root remain open, or should I close them once the SOFIE-side PRs are submitted?

Thanks,
Aditya

**Comment by AdityaDRathore** [comment]
Hi @dpiparo, 

I looked into re-opening this on the SOFIE dev repository, but the Softplus operator (kSoftplus in `ROperator_BasicUnary`) does not yet exist in the standalone SOFIE repo, it was added directly to root-project/root. 

This fix would need to be applied here in ROOT, or the Softplus operator would need to be ported to ML4EP/SOFIE first. 

Please let me know how you'd like to proceed.
